src/Guide/Api/Utils.hs

```diff
@@ -89,3 +103,61 @@ instance (HasSwagger api, KnownSymbol name, KnownSymbol desc) =>
              symbolVal (Proxy @desc))
             Nothing
      in toSwagger (Proxy @api) & applyTags [tag]
+
+-- | Servant request details, can be captured by adding @RequestDetails :>@ to an API branch.
+data RequestDetails = RequestDetails
+  { rdIp        :: Maybe IP    -- ^ request ip address
+  , rdReferer   :: Maybe Text  -- ^ request referrer
+  , rdUserAgent :: Maybe Text  -- ^ request User-Agent
+  } deriving (Show, Generic)
+
+instance FromHttpApiData IP where
+  parseUrlPiece text = either (Left . toText) Right $ readEither $ toString text
+
+instance (HasSwagger api) => HasSwagger (RequestDetails :> api) where
+  toSwagger _ = toSwagger (Proxy :: Proxy api)
+
+instance (HasServer api context)
+  => HasServer (RequestDetails :> api) context where
+
+  type ServerT (RequestDetails :> api) m = RequestDetails -> ServerT api m
+
+  hoistServerWithContext _ pc nt s = hoistServerWithContext (Proxy :: Proxy api) pc nt . s
+
+  route Proxy context subserver = route (Proxy :: Proxy api) context $
+      subserver `addHeaderCheck` withRequest getRequestDetails
+    where
+      getRequestDetails :: Request -> DelayedIO RequestDetails
+      getRequestDetails req = pure $ RequestDetails mIp mReferer mUseragent
+        where
+          mIp :: Maybe IP
+          mIp =  getIp $ lookupName "Forwarded-For" <|> lookupName "X-Forwarded-For"
+          mReferer :: Maybe Text
+          mReferer = getHeader hReferer
+          mUseragent :: Maybe Text
+          mUseragent = getHeader hUserAgent
+
+          lookupName :: NHTH.HeaderName -> Maybe ByteString
+          lookupName headerName = lookup headerName (requestHeaders req)
+
+          getHeader :: FromHttpApiData a => NHTH.HeaderName -> Maybe a
+          getHeader headerName = join $
+            (either (\_ -> Nothing) Just) . parseHeader <$> lookupName headerName
+
+          getIp :: Maybe ByteString -> Maybe IP
+          getIp mBody = case mBody of
+            Nothing -> sockAddrToIP $ remoteHost req
+            Just ff -> case readMaybe (toString ip) of
+              Nothing -> error $ "couldn't read Forwarded-For address: " ++
```

Comment:
Oh, this shouldn't error out – it should return `Nothing`. Yeah, my original code errored out, but that was a mistake.